### PR TITLE
lib: Add nghttp2_hd_deflate_hd_vect() deflate API to support multiple…

### PR DESCRIPTION
**Pull request by weliu**
… bufs input

### CODE REVIEW

File: lib/nghttp2_buf.c [hunk 1]
```diff
@@ -223,13 +223,60 @@ int nghttp2_bufs_wrap_init(nghttp2_bufs *bufs, uint8_t *begin, size_t len,
...
+  nghttp2_buf_chain *head_chain = NULL;
+
+  for (i = 0; i < in_len; ++i) {
+    uint8_t *begin = bufs_in[i];
+
+    cur_chain = nghttp2_mem_malloc(mem, sizeof(nghttp2_buf_chain));
+    if (cur_chain == NULL) {
+      while (head_chain) {
+        nghttp2_mem_free(mem, head_chain);
+        head_chain = head_chain->next;
```
Comment (tatsuhiro-t) [inline lib/nghttp2_buf.c]: head_chain was already freed.  Accessing head_chain->next is heap-use-after-free.

File: lib/nghttp2_buf.c [hunk 2]
```diff
@@ -223,13 +223,60 @@ int nghttp2_bufs_wrap_init(nghttp2_bufs *bufs, uint8_t *begin, size_t len,
...
 void nghttp2_bufs_wrap_free(nghttp2_bufs *bufs) {
   if (bufs == NULL) {
     return;
   }
 
-  nghttp2_mem_free(bufs->mem, bufs->head);
-  bufs->head = NULL;
+  while (bufs->head) {
+    nghttp2_mem_free(bufs->mem, bufs->head);
+    bufs->head = bufs->head->next;
```
Comment (tatsuhiro-t) [inline lib/nghttp2_buf.c]: bufs->head was already freed.  Accessing bufs->head->next is heap-use-after-free.

File: lib/nghttp2_buf.h [hunk 3]
```diff
@@ -211,6 +211,22 @@ int nghttp2_bufs_wrap_init(nghttp2_bufs *bufs, uint8_t *begin, size_t len,
                            nghttp2_mem *mem);
 
 /*
+ * Initializes |bufs| using supplied |in_len| size of buf vector |bufs_in|,
+ * each buf has length |buf_len|.
+ * The buffer size is fixed and no allocate extra chunk buffer is allocated.
```
Comment (tatsuhiro-t) [inline lib/nghttp2_buf.h]: Probably it should read as "no extra chunk buffer is allocated."

File: lib/nghttp2_hd.c [hunk 4]
```diff
@@ -1503,6 +1503,38 @@ ssize_t nghttp2_hd_deflate_hd(nghttp2_hd_deflater *deflater, uint8_t *buf,
   return (ssize_t)buflen;
 }
 
+ssize_t nghttp2_hd_deflate_hd_vect(nghttp2_hd_deflater *deflater, uint8_t **bufsin,
+                              size_t inlen, size_t buflen, const nghttp2_nv *nv,
+                              size_t nvlen) {
```
Comment (tatsuhiro-t) [inline lib/nghttp2_hd.c]: We have standard struct io_vec, so perhaps, nghttp2_hd_deflate_hd_vec() sounds more like C?

File: lib/includes/nghttp2/nghttp2.h [hunk 5]
```diff
@@ -4519,6 +4519,38 @@ nghttp2_hd_deflate_hd(nghttp2_hd_deflater *deflater, uint8_t *buf,
 /**
  * @function
  *
+ * Deflates the |nva|, which has the |nvlen| name/value pairs, into
+ * the |inlen| size of buf vector |bufsin|, each buf has length |buflen|.
+ *
```
Comment (tatsuhiro-t) [inline lib/includes/nghttp2/nghttp2.h]: It would be nice to tell the reader that if one chunk is used up, next chunk is used.

**Comment by tatsuhiro-t** [comment]
One thing we have to work out is that nghttp2 deflate code does not necessarily use whole chunk before advancing to the next chunk.
Example: https://github.com/nghttp2/nghttp2/blob/8579b8a968226e7aa23def52effe0e968c9aedb5/lib/nghttp2_hd_huffman.c#L67

So we need some output parameter to tell the caller that how many bytes are used per chunk.

File: lib/nghttp2_buf.c (hunk 1 above)
Comment (weliu) [inline lib/nghttp2_buf.c]: You are totally right, fixed it in latest PR.

File: lib/nghttp2_buf.c (hunk 2 above)
Comment (weliu) [inline lib/nghttp2_buf.c]: Same as above.

File: lib/nghttp2_buf.h (hunk 3 above)
Comment (weliu) [inline lib/nghttp2_buf.h]: Fixed.

File: lib/nghttp2_hd.c (hunk 4 above)
Comment (weliu) [inline lib/nghttp2_hd.c]: Yes, that is more reasonable.

File: lib/includes/nghttp2/nghttp2.h (hunk 5 above)
Comment (weliu) [inline lib/includes/nghttp2/nghttp2.h]: Fixed.

**Comment by tatsuhiro-t** [comment]
Thanking about the code above, it indicates that we require at least 4 bytes per chunk.

**Comment by weliu** [comment]
I have added a new function nghttp2_bufs_len_vec() in nghttp2_buf.c to output each size of buf. Thanks for your advice.

**Comment by tatsuhiro-t** [comment]
Thank you.
As for the current nghttp2 implementation which might not fill up all chunk, I will see I can fix it so that it always uses all chunk provided.
Other than that, PR is good to go.

**Comment by tatsuhiro-t** [comment]
Merged with some amends. Thank you!

**Pull request merged**